### ISSUE TITLE
Installation issue on AWS

### ISSUE THREAD

**Issue by chaitanyamalaviya** [clab/dynet#282]
I am trying to get dynet-gpu working on an AWS machine using the Deep Learning AMI image. While running make, I face an error when the dynet examples are being linked. After some searching, it seems that the issue may be resolved by adding -lboost-serialization flag to g++. I am using boost version 1.62.0.
I haven't figured out how to fix this yet though. Here is the error log, which is repeated while linking all the examples:

```
Linking CXX executable 1_linear_regression
../dynet/libdynet.so: undefined reference to `boost::archive::text_iarchive_impl<boost::archive::text_iarchive>::load_override(boost::archive::class_name_type&)'
../dynet/libdynet.so: undefined reference to `boost::archive::basic_binary_iarchive<boost::archive::binary_iarchive>::load_override(boost::archive::class_name_type&)'
../dynet/libdynet.so: undefined reference to `boost::archive::archive_exception::archive_exception(boost::archive::archive_exception const&)'
../dynet/libdynet.so: undefined reference to `boost::serialization::singleton_module::is_locked()'
collect2: error: ld returned 1 exit status
make[2]: *** [tutorial/0_multiply] Error 1
make[1]: *** [tutorial/CMakeFiles/0_multiply.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by neubig** [comment]
Can you try to follow this and send your `make.log` file? http://dynet.readthedocs.io/en/latest/install.html#debugging-build-problems

**Comment by chaitanyamalaviya** [comment]
Here is the make.log file:
[make.log.txt](https://github.com/clab/dynet/files/752015/make.log.txt)

**Comment by neubig** [comment]
@chaitanyamalaviya I'm not sure, but I suspect that this might be a problem that is caused by finicky versions of gcc that require a particular ordering of libraries in the command line (you must position a library earlier in the command line than the libraries it uses). I made a new branch that could potentially fix the problem here: https://github.com/clab/dynet/tree/fix-cuda-linker (the `fix-cuda-linker` branch). Could you try to check this out and compile?

**Comment by chaitanyamalaviya** [comment]
Thanks for taking a look! I tried the above but the same error persisted..

**Comment by pmichel31415** [comment]
FYI I encounter the same problem on my machine (and the fix didn't work).

The following worked however (with cuda 7.5) : 

   CC=gcc-4.9 CXX=g++-5 cmake .. -DEIGEN3_INCLUDE_DIR=/path/to/eigen -DBACKEND=cuda

TL;DR : use gcc 4.9 but g++ 5, also compile your programs with g++-5

Hope that helps

**Comment by neubig** [comment]
I was able to reproduce this on the AWS Deep Learning AMI, so I'll try to debug when I have another moment.

**Comment by neubig** [comment]
So the problem appears to be that the deep learning AMI has a broken Boost installation. Two versions of the Boost libraries installed (1.53 in /usr/lib64, and 1.62 in /usr/lib), and when DyNet compiles it was finding the headers from one version 1.62, and the libraries from another 1.53. By removing the libraries from 1.53 I was able to fix the problem.

    mkdir ~/boost-quarantine
    sudo mv /usr/lib64/libboost* ~/boost-quarantine

Could you try this? For the short-term this work-around should work, and should not give you segfaults, although it may break other programs. In the long-term, I really want to get rid of the dependency on boost :)

**Comment by xunzhang** [comment]
🍺 I also agree to get rid of boost, I am working on dynet-84.

**Comment by chaitanyamalaviya** [comment]
I can confirm that the above fix lets dynet-gpu compile successfully. I tried running some of the python examples from the /examples/python directory. I allocate 2GB of memory to dynet. While the examples do not crash during execution, I receive a python error just before the execution terminates, followed by a Backtrace and a Memory Map.
`*** Error in `python': double free or corruption (!prev): 0x0000000001a06960 ***`

**Comment by neubig** [comment]
Thanks for the confirmation! The double memory free is another issue, so I'll close this one: https://github.com/clab/dynet/issues/310
